devnet-5 `proposer_preferences` exception

we probably don't need to print a warning here, assuming the check is right

```
2026-06-07 07:29:47.183 WARN  - Encountered exception while processing message for topic /eth2/a4da394d/proposer_preferences/ssz_snappy
java.util.concurrent.CompletionException: tech.pegasys.teku.spec.datastructures.forkchoice.InvalidCheckpointException: Checkpoint state (19948) must be at or prior to checkpoint slot boundary (19936)
at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.uniApplyNow(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.uniApplyStage(Unknown Source)
at java.base/java.util.concurrent.CompletableFuture.thenApply(Unknown Source)
at tech.pegasys.teku.infrastructure.async.SafeFuture.thenApply(SafeFuture.java:603)
at tech.pegasys.teku.dataproviders.generators.StateAtSlotTask.performTask(StateAtSlotTask.java:96)
at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74)
at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$0(ScheduledExecutorAsyncRunner.java:124)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
at java.base/java.lang.Thread.run(Unknown Source)
Caused by: tech.pegasys.teku.spec.datastructures.forkchoice.InvalidCheckpointException: Checkpoint state (19948) must be at or prior to checkpoint slot boundary (19936)
at tech.pegasys.teku.dataproviders.generators.StateAtSlotTask.regenerateFromState(StateAtSlotTask.java:108)
at java.base/java.util.Optional.map(Unknown Source)
at tech.pegasys.teku.dataproviders.generators.StateAtSlotTask.lambda$performTask$0(StateAtSlotTask.java:96)
... 10 more
```

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidator.java
@@ -121,6 +121,12 @@ public class ProposerPreferencesGossipValidator {
     final int minSeedLookahead = spec.atSlot(proposalSlot).getConfig().getMinSeedLookahead();
     final UInt64 checkpointEpoch =
         spec.computeEpochAtSlot(proposalSlot).minusMinZero(minSeedLookahead);
+
+    /*
+     * Pre-flight the checkpoint-state precondition for the [REJECT] is_valid_proposal_slot rule below.
+     * A dependent root at or after the checkpoint boundary cannot be the root of the checkpoint state
+     * required by that rule.
+     */
     final UInt64 checkpointBoundarySlot = spec.computeStartSlotAtEpoch(checkpointEpoch);
     final Optional<UInt64> maybeDependentRootSlot =
         recentChainData.getSlotForBlockRoot(dependentRoot);

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidator.java
@@ -19,6 +19,7 @@ import static tech.pegasys.teku.statetransition.validation.InternalValidationRes
 import static tech.pegasys.teku.statetransition.validation.InternalValidationResult.ignore;
 import static tech.pegasys.teku.statetransition.validation.InternalValidationResult.reject;
 
+import java.util.Optional;
 import java.util.Set;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
@@ -120,6 +121,23 @@ public class ProposerPreferencesGossipValidator {
     final int minSeedLookahead = spec.atSlot(proposalSlot).getConfig().getMinSeedLookahead();
     final UInt64 checkpointEpoch =
         spec.computeEpochAtSlot(proposalSlot).minusMinZero(minSeedLookahead);
+    final UInt64 checkpointBoundarySlot = spec.computeStartSlotAtEpoch(checkpointEpoch);
+    final Optional<UInt64> maybeDependentRootSlot =
+        recentChainData.getSlotForBlockRoot(dependentRoot);
+    if (maybeDependentRootSlot.isPresent()) {
+      final UInt64 dependentRootSlot = maybeDependentRootSlot.get();
+      if (!dependentRootSlot.isLessThan(checkpointBoundarySlot)) {
+        LOG.trace(
+            "Proposer preferences dependent root {} is at slot {}, but must be before checkpoint boundary slot {}",
+            dependentRoot,
+            dependentRootSlot,
+            checkpointBoundarySlot);
+        return completedFuture(
+            reject(
+                "Proposer preferences dependent root %s is at slot %s, but must be before checkpoint boundary slot %s",
+                dependentRoot, dependentRootSlot, checkpointBoundarySlot));
+      }
+    }
     return recentChainData
         .retrieveCheckpointState(new Checkpoint(checkpointEpoch, dependentRoot))
         .thenApply(
@@ -134,7 +152,9 @@ public class ProposerPreferencesGossipValidator {
               final BeaconState state = maybeState.get();
 
               /*
-               * [REJECT] is_valid_proposal_slot(state, preferences) returns True
+               * [REJECT] is_valid_proposal_slot(state, preferences) returns True, where state is the checkpoint state
+               * at the epoch compute_epoch_at_slot(preferences.proposal_slot) - MIN_SEED_LOOKAHEAD
+               * and the root preferences.dependent_root.
                */
               final int slotsPerEpoch = spec.atSlot(proposalSlot).getConfig().getSlotsPerEpoch();
               final int lookaheadIndex =
@@ -166,6 +186,17 @@ public class ProposerPreferencesGossipValidator {
               }
 
               return ACCEPT;
+            })
+        .exceptionally(
+            error -> {
+              LOG.trace(
+                  "Unable to generate proposer preferences checkpoint state for checkpoint epoch {} and dependent root {}",
+                  checkpointEpoch,
+                  dependentRoot,
+                  error);
+              return reject(
+                  "Unable to generate proposer preferences checkpoint state for checkpoint epoch %s and dependent root %s",
+                  checkpointEpoch, dependentRoot);
             });
   }
 

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidatorTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidatorTest.java
@@ -125,6 +125,45 @@ public class ProposerPreferencesGossipValidatorTest {
   }
 
   @TestTemplate
+  void shouldReject_whenDependentRootBlockIsNotBeforeCheckpointBoundary() {
+    final UInt64 checkpointBoundarySlot = getCheckpointBoundarySlot();
+    when(recentChainData.getSlotForBlockRoot(dependentRoot))
+        .thenReturn(Optional.of(checkpointBoundarySlot));
+
+    assertThatSafeFuture(validator.validate(signedProposerPreferences))
+        .isCompletedWithValueMatching(
+            result ->
+                result.isReject()
+                    && result
+                        .getDescription()
+                        .filter(
+                            description ->
+                                description.contains("but must be before checkpoint boundary slot"))
+                        .isPresent());
+    verify(recentChainData, never()).retrieveCheckpointState(any(Checkpoint.class));
+  }
+
+  @TestTemplate
+  void shouldReject_whenCheckpointStateRetrievalFails() {
+    when(recentChainData.retrieveCheckpointState(any(Checkpoint.class)))
+        .thenReturn(SafeFuture.failedFuture(new IllegalStateException("checkpoint state failed")));
+
+    assertThatSafeFuture(validator.validate(signedProposerPreferences))
+        .isCompletedWithValueMatching(
+            result ->
+                result.isReject()
+                    && result
+                        .getDescription()
+                        .filter(
+                            description ->
+                                description.contains(
+                                    "Unable to generate proposer preferences checkpoint state"))
+                        .isPresent());
+    verify(gossipValidationHelper, never())
+        .isSignatureValidWithRespectToProposerIndex(any(), any(), any(), any());
+  }
+
+  @TestTemplate
   void shouldIgnore_whenAlreadySeen() {
     assertThatSafeFuture(validator.validate(signedProposerPreferences))
         .isCompletedWithValue(ACCEPT);
@@ -220,5 +259,11 @@ public class ProposerPreferencesGossipValidatorTest {
 
     return signedProposerPreferencesSchema.create(
         proposerPreferences, dataStructureUtil.randomSignature());
+  }
+
+  private UInt64 getCheckpointBoundarySlot() {
+    return spec.computeStartSlotAtEpoch(
+        spec.computeEpochAtSlot(proposalSlot)
+            .minusMinZero(spec.atSlot(proposalSlot).getConfig().getMinSeedLookahead()));
   }
 }

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidatorTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/ProposerPreferencesGossipValidatorTest.java
@@ -126,7 +126,10 @@ public class ProposerPreferencesGossipValidatorTest {
 
   @TestTemplate
   void shouldReject_whenDependentRootBlockIsNotBeforeCheckpointBoundary() {
-    final UInt64 checkpointBoundarySlot = getCheckpointBoundarySlot();
+    final UInt64 checkpointBoundarySlot =
+        spec.computeStartSlotAtEpoch(
+            spec.computeEpochAtSlot(proposalSlot)
+                .minusMinZero(spec.atSlot(proposalSlot).getConfig().getMinSeedLookahead()));
     when(recentChainData.getSlotForBlockRoot(dependentRoot))
         .thenReturn(Optional.of(checkpointBoundarySlot));
 
@@ -259,11 +262,5 @@ public class ProposerPreferencesGossipValidatorTest {
 
     return signedProposerPreferencesSchema.create(
         proposerPreferences, dataStructureUtil.randomSignature());
-  }
-
-  private UInt64 getCheckpointBoundarySlot() {
-    return spec.computeStartSlotAtEpoch(
-        spec.computeEpochAtSlot(proposalSlot)
-            .minusMinZero(spec.atSlot(proposalSlot).getConfig().getMinSeedLookahead()));
   }
 }

